### DISCUSSION TITLE
LastDayOfMonth

### DISCUSSION THREAD

**Discussion by luckman212** [SilentVoid13/Templater#139] [Templates Showcase]
**Update:** see [below](https://github.com/SilentVoid13/Templater/discussions/139#discussioncomment-646175) for the simpler more elegant solution. Leaving this here though in case some of the other methods used are useful for others.

Will insert the last day of the current month.
You can also select a date in the editor and then it will use that as the basis for the calculation

Related discussion: https://github.com/SilentVoid13/Templater/issues/93

```js
<%*
function pad(n) {
  return n.toString().padStart(2, "0");
}
function fmtDate(d) {
  var yyyy = d.getFullYear();
  var mm = pad(d.getMonth() + 1);
  var dd = pad(d.getDate());
  var dArr = [ yyyy, mm, dd ];
  return(dArr.join('-'));
}
function lastDay(a) {
  if (typeof(a) === 'undefined' || a === '') {
    var d = new Date();
  } else {
    var d = new Date(a + ' 12:00:00');
    if (d.toString() === 'Invalid Date') {
      console.log('invalid date');
      return
    }
  }
  var lastDay = new Date(d.getFullYear(),d.getMonth()+1,0);
  lastDate = fmtDate(lastDay)
  lastDay = lastDay.getDate()
  return lastDate // change to "lastDay" to output day only
} %><% lastDay(tp.file.selection())%>
```

**Comment by SilentVoid13** [comment]
Actually for the first / last day of the current month, you can just do the following, since the `moment` object is available in the global namespace:

```js
<%* tR += moment().startOf('month').format('YYYY-MM-DD') %>
<%* tR += moment().endOf('month').format('YYYY-MM-DD') %>
```

You could tweak this to use `tp.file.selection()` like you did. Using `moment` when it comes to dates usually makes things easier!

**Reply by luckman212** [reply]
Yes that sure is a lot easier! Thanks SV. Edited my original post to point to your moment.js solution.

**Reply by ghost** [reply]
Hi @SilentVoid13,
This is probably a stupid question but i'm new to templater:

You use "tR" in above code snippet. What is the meaning of that? It's not a user defined variable because I changed it to "x" and the code didn't work anymore. Is it a system variable or something alike?

Thanks in advance

**Reply by serpro69** [reply]
@SamAdams51 https://silentvoid13.github.io/Templater/commands/execution-command.html#how-to-output-a-value-from-a-javascript-execution-command-

**Reply by ghost** [reply]
Hi @serpro69,
thanks for pointing me in the right direction. It's quite clear now.
Thanks again and have a nice day.
